### PR TITLE
Feat: implements product search constraints

### PR DESCRIPTION
**Pull request by SWH-Relewise**

### CODE REVIEW

File: packages/client/src/builders/search/searchConstraintBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+            $type: 'Relewise.Client.Requests.Search.Settings.ResultMustHaveVariantConstraint, Relewise.Client',
+            ...constaint,
+        } as ResultMustHaveVariantConstraint;
+        return this;
+    }
+
+    build(): ResultMustHaveVariantConstraint | null {
+        // Do to how the data contracts are generated, resultsConstraints on searchSettings expect this specific type
+        // Once more types are added it will expect the generic type and this cast should be removed and the generic type should be returned
+        return this.resultConstraint as ResultMustHaveVariantConstraint | null;
```
Comment (mzanoni) [inline packages/client/src/builders/search/searchConstraintBuilder.ts:17]: Really good comment!

File: packages/client/tests/unit-tests/builders/search/productSearchBuilder.unit.test.ts [hunk 2]
```diff
@@ -94,4 +94,14 @@ test('variantSearchSettings', () => {
         .build();
 
     expect(subject.settings?.variantSettings?.excludeResultsWithoutVariant).toBe(true);
+});
+
+test('resultMustHaveVariantConstraint', () => {
+    const subject: ProductSearchRequest = baseBuilder()
+        .searchConstraints(s => s.setResultMustHaveVariantConstraint({
+            exceptWhenProductHasNoVariants: true,
```
Comment (mzanoni) [inline packages/client/tests/unit-tests/builders/search/productSearchBuilder.unit.test.ts:102]: 😍

File: packages/client/src/builders/search/searchConstraintBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { ProductSearchResultConstraint, ResultMustHaveVariantConstraint } from 'src/models/data-contracts';
```
Comment (mzanoni) [inline packages/client/src/builders/search/searchConstraintBuilder.ts:1]: Import is wrong here :)

**Review by mzanoni — CHANGES_REQUESTED** [review]
Awesome :)

Be aware that the build is broken atm

File: packages/client/src/builders/search/searchConstraintBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+export class SearchConstraintBuilder {
+    private resultConstraint: ProductSearchResultConstraint | null = null;
+
+    public setResultMustHaveVariantConstraint(constaint: { exceptWhenProductHasNoVariants: boolean }): this {
+        this.resultConstraint = {
+            $type: 'Relewise.Client.Requests.Search.Settings.ResultMustHaveVariantConstraint, Relewise.Client',
+            ...constaint,
+        } as ResultMustHaveVariantConstraint;
```
[the remark this answers is not in the document]
Comment (SWH-Relewise) [reply packages/client/src/builders/search/searchConstraintBuilder.ts:10]: This syntax is invalid here! 
image

[the remark this answers is not in the document]
Comment (mzanoni) [reply packages/client/src/builders/search/searchConstraintBuilder.ts:10]: Just assign it to as const before assigning to `this.resultConstraint`

[the remark this answers is not in the document]
Comment (SWH-Relewise) [reply packages/client/src/builders/search/searchConstraintBuilder.ts]: Will do! 🫱🪖

[the remark this answers is not in the document]
Comment (SWH-Relewise) [reply packages/client/src/builders/search/searchConstraintBuilder.ts]: There you go!

**Review by mzanoni — APPROVED** [review]

**Pull request merged**